torch/csrc/distributed/rpc/python_rpc_handler.cpp

```diff
@@ -3,43 +3,41 @@
 namespace torch {
 namespace distributed {
 namespace rpc {
-namespace {
-  py::object module_;
-  py::object runUDFFunction_;
-  py::object loadResultFunction_;
-} // anonymous namespace
 
-namespace PythonRpcHandler {
-  void init() {
-    AutoGIL ag;
-    if (module_ == nullptr) {
-      module_ = py::module::import("torch.distributed.internal_rpc_utils");
-    }
-    if (runUDFFunction_ == nullptr) {
-      runUDFFunction_ = module_.attr("run_python_udf_internal");
-    }
-    if (loadResultFunction_ == nullptr) {
-      loadResultFunction_ = module_.attr("load_python_udf_result_internal");
-    }
+PythonRpcHandler::PythonRpcHandler() {
+  AutoGIL ag;
+  py::object module =
+      py::module::import("torch.distributed.internal_rpc_utils");
+  if (runUDFFunction_ == nullptr) {
+    runUDFFunction_ = module.attr("run_python_udf_internal");
   }
-
-  std::vector<char> generatePythonUDFResult(
-    const Message& request) {
-    AutoGIL ag;
-    auto pargs = py::bytes(request.payload().data(), request.payload().size());
-    py::bytes pres = runUDFFunction_(pargs);
-    const auto& presStr = static_cast<std::string>(pres);
-    std::vector<char> payload(presStr.begin(), presStr.end());
-    return payload;
+  if (loadResultFunction_ == nullptr) {
+    loadResultFunction_ = module.attr("load_python_udf_result_internal");
   }
+}
 
-  py::object loadPythonUDFResult(const Message& message) {
-    AutoGIL ag;
-    auto pargs = py::bytes(message.payload().data(), message.payload().size());
-    return loadResultFunction_(pargs);
-  }
-} // PythonRpcHandler
+PythonRpcHandler& PythonRpcHandler::getInstance() {
+  static PythonRpcHandler handler;
+  return handler;
+}
+
+std::vector<char> PythonRpcHandler::generatePythonUDFResult(
+    const Message& request) {
+  AutoGIL ag;
+  auto pargs = py::bytes(request.payload().data(), request.payload().size());
+  TORCH_CHECK(runUDFFunction_ != nullptr, "runUDFFunction_ is nullptr");
+  py::bytes pres = runUDFFunction_(pargs);
```

Comment:
IIUC, only `runUDFFunction_` needs to be protected by GIL? Can we move the rest of the code outside GIL?